Enhance word boundary for #136

k94-ishi opened the pull request:
closes #136

This PR enhances the handling of `\b` word boundaries.

## Added syntax

- `\b{start}`: Left word boundary (equivalent to `\<`)
- `\b{end}`: Right word boundary (equivalent to `\>`)
- `\b{start-half}`: Left word half boundary
- `\b{end-half}`: Right word half boundary

## Implementation details

- Extended the parser to support brace-style word boundary cases
- Added new `Assertion` variants to represent the half-boundary semantics
  - `Assertion::LeftWordHalfBoundary`
  - `Assertion::RightWordHalfBoundary`

## Notes

- Provides clear error messages for unsupported `\B{...}` syntaxes
- Adds tests for boundary brace parsing and matching
  - Parsing validation tests for valid and invalid syntax
  - Functional tests to verify boundary semantics

keith-hall reviewed (commented):
Thanks, looks good. I have a few small ideas for improvement.

CAD97 commented on src/parse.rs:
```diff
@@ -689,6 +681,62 @@ impl<'a> Parser<'a> {
+
+        // Find the opening brace (should be at ix + 2, after \b or \B)
+        let start_brace = ix + 2;
+        if bytes.get(start_brace) != Some(&b'{') {
+            return Err(Error::ParseError(
+                ix,
+                ParseError::InvalidEscape("\\b{...}".to_string()),
+            ));
+        }
```
The usual pattern is to `assert!` for things that should always be true by construction, but since emitting a proper error here is nearly as trivial as the panic, doing so seems reasonable.

However, only the `{` and not the `\b` is checked currently, which feels a little odd.

Assert:

```rust
        assert_matches!(
            bytes.get(ix..ix + 3),
            Some([b'\\', b'b' | 'B', b'{']),
            "ix should be at \\b{{...}}",
        );
```

Error:

```
        if !matches!(bytes.get(ix..ix + 3), Some([b'\\', b'b' | 'B', b'{'])) {
            return Err(Error::ParseError(
                ix,
                ParseError::InvalidEscape("\\b{...}".to_string()),
            );
        }
```

keith-hall commented on tests/parse.rs:
Just curious, under what circumstances do we prefer having a separate test file compared to having the test be a part of the unit being tested? i.e. https://github.com/fancy-regex/fancy-regex/blob/main/src/parse.rs#L1161 ? Like having separate tests, we use public API I guess, having them in the module being tested can use private APIs and get closer to the raw details...

keith-hall commented on tests/parse.rs:
```diff
@@ -0,0 +1,61 @@
+#[test]
+fn word_boundary_brace_parsing_ok() {
+    // Valid patterns should succeed
+    assert!(Regex::new(r"\b{start}").is_ok());
+    assert!(Regex::new(r"\b{end}").is_ok());
+    assert!(Regex::new(r"\b{start-half}").is_ok());
+    assert!(Regex::new(r"\b{end-half}").is_ok());
```
if these were in the parser itself's `test` (sub)module as linked above, we could assert that parsing `\b{start}` produces an `Expr::Assertion(Assertion::LeftWordBoundary)` as opposed to just that constructing a regex with this pattern succeeds. I feel like it would be more beneficial, what do others think?

keith-hall reviewed (commented):
Thanks for the updates. I have a couple of more questions/considerations - this may be where my lack of Rust experience shows, so please feel free to tell me that the way it is done now is better than my proposal :)

keith-hall commented:
One more quick thing, in our [contributing guide](https://github.com/fancy-regex/fancy-regex/blob/main/CONTRIBUTING.md#contributing) we mention keeping the documentation comments in lib.rs up to date with new syntax. It would be great to see that as well please 😎

k94-ishi commented:
I've moved the tests that were in `tests/parse.rs` into corresponding unit tests in `src/parse.rs`, and then removed `tests/parse.rs`. Since parsing is just one internal step of regex processing, it fits better as a unit test target rather than an integration test.
As for the docs in lib.rs, I checked and since this syntax already exists in the regex crate, I think no update is needed — but happy to add one if you prefer.

keith-hall reviewed (approved):
Looks great, thank you!

k94-ishi commented:
These syntaxes were requested in #136, and they come from the Rust `regex` crate — they’re not specific to Oniguruma.
[https://docs.rs/regex/1.11.3/regex/#empty-matches](https://docs.rs/regex/1.11.3/regex/#empty-matches:~:text=exactly%20n%20x-,%C2%A7,-Empty%20matches)

Since the existing documentation in `lib.rs` already mentions that fancy-regex supports the `regex` crate’s syntax, I think no update is needed.
But I can add a note if needed.

k94-ishi commented:
I've implemented the changes based on your feedback. Here’s a summary of the updates:

1. Updated `is_repeatable()` to check Oniguruma flag:
   - Default mode: assertions can be repeated (e.g. `\b{1}`, `\B{2}`, `^*`, etc.)
   - Oniguruma mode: returns error "Target of repeat operator is invalid"
2. Enhanced `\b{...}` parsing to distinguish between:
   - Repetition: `\b{1}` - checks for digits or commas after `{`
   - Named boundaries: \b{start} - continues to work as before
3. Updated tests:
   - Added tests for both modes
   - Removed old tests that expected assertion repetition to be an error

Please let me know if anything looks off or could be improved!

Note: The slight coverage decrease is expected, as some old tests that previously expected assertion repetition to be an error were removed.

keith-hall commented:
> Note: The slight coverage decrease is expected, as some old tests that previously expected assertion repetition to be an error were removed.

Could we not keep them with the Oniguruma mode flag? :slightly_smiling_face: Sometimes unit tests can be really helpful, those were helping to validate that the error location reporting was accurate, that each different repetition type is picked up etc.

I think we need a few small tweaks for "extended" mode, the regex crate seems to strip whitespace inside the curly braces, but fancy-regex in this PR currently reports an error.
```regex
(?x)\b{ s t a r t}\w
```
<img width="662" height="369" alt="[image](https://rustexp.lpil.uk/)" src="https://github.com/user-attachments/assets/1b2f27a7-e0c2-4a59-a658-fd6841df5b67" />
<img width="1163" height="436" alt="fancy-regex parse error" src="https://github.com/user-attachments/assets/f0bf2fe4-8139-4fca-947b-d29ae4f0cfeb" />

k94-ishi commented:
You’re absolutely right — I’ve reintroduced those tests under the Oniguruma mode.  

I also fixed the behavior under `FLAG_IGNORE_SPACE` (extended mode):  
`\b{...}`, `\x{...}`, `\u{...}`, and `\U{...}` now correctly skip whitespace and comments inside braces, matching the `regex` crate’s behavior.  

`(?x: \\b { s t a r t })` → parsed as `\\b{start}`  
`(?x: \\b { s t a r t # \n})` → parsed as `\\b{start}`  
`(?x: \\u { 0 0 3 2 })` → parsed as `\\u{0032}`  
`(?x: \\u { 0 0 3 2 # \n })` → parsed as `\\u{0032}`  

I used the function `optional_whitespace` to handle this, so some syntaxes behave slightly differently from the `regex` crate.  

`(?x: \\b { s t a r t (?#)})` → parsed as `\\b{start}`  

Please take a look when you get a chance!

k94-ishi commented:
In the `regex` crate’s behavior,

- `(?x: \\b { s t a r t })`

```plaintext
Syntax(
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
regex parse error:
    (?x: \b { s t a r t })
              ^
error: repetition quantifier expects a valid decimal
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
)
```

- `(?x: \\b{ s t a r t (?#)})`

``` plaintext
Syntax(
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
regex parse error:
    (?x: \b{ s t a r t (?#)})
           ^^^^^^^^^^^^
error: special word boundary assertion is either unclosed or contains an invalid character
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
)
```

keith-hall reviewed (approved):
Awesome, thanks very much for your hard work on this one!

keith-hall commented:
> If these cases should instead raise errors to match the `regex` crate’s behavior, I’ll update them accordingly.

I think trying to exactly match another engine's behavior in all circumstances is an impossible goal. I'm happy with how it is now - the important thing is that all constructs which the regex crate allows are also allowed by fancy-regex when not in Oniguruma mode, for compatibility. If we don't error when regex does (i.e. fancy-regex is more lenient), I think it is fine.
